subsys/fs/zms/zms.c

```diff
@@ -0,0 +1,1662 @@
+	   - This ATE was never overwritten in the next 255 cycles because of
+	     large data size
+	   - Next 256th cycle the leading cycle_cnt is 0, this ATE becomes
+	     valid even if it is not the case.
+	 */
+	close_ate.cycle_cnt = fs->sector_cycle;
+	memset(&garbage_ate, fs->flash_parameters->erase_value, sizeof(garbage_ate));
+	while (fs->ate_wra >= (fs->data_wra + ate_size)) {
+		rc = zms_flash_ate_wrt(fs, &garbage_ate);
+		if (rc) {
+			return rc;
+		}
+	}
+
+	fs->ate_wra &= ADDR_SECT_MASK;
+	fs->ate_wra += (fs->sector_size - ate_size);
+
+	zms_ate_crc8_update(&close_ate);
+
+	(void)zms_flash_ate_wrt(fs, &close_ate);
+
+	/* skip the empty ATE */
+	fs->ate_wra -= ate_size;
+
+	zms_sector_advance(fs, &fs->ate_wra);
+
+	rc = zms_get_sector_cycle(fs, fs->ate_wra, &fs->sector_cycle);
+	if (rc == -ENOENT) {
+		/* sector never used */
+		fs->sector_cycle = 0;
+	} else if (rc) {
+		/* bad flash read */
+		return rc;
+	}
+
+	fs->data_wra = fs->ate_wra & ADDR_SECT_MASK;
+
+	return 0;
+}
+
+static int zms_add_gc_done_ate(struct zms_fs *fs)
+{
+	struct zms_ate gc_done_ate;
+
+	LOG_DBG("Adding gc done ate at %llx", fs->ate_wra);
+	gc_done_ate.id = ZMS_HEAD_ID;
+	gc_done_ate.len = 0U;
+	gc_done_ate.offset = (uint32_t)(fs->data_wra & ADDR_OFFS_MASK);
+	gc_done_ate.metadata = 0xffffffff;
+	gc_done_ate.cycle_cnt = fs->sector_cycle;
+
+	zms_ate_crc8_update(&gc_done_ate);
+
+	return zms_flash_ate_wrt(fs, &gc_done_ate);
+}
+
+static int zms_add_empty_ate(struct zms_fs *fs, uint64_t addr)
+{
+	struct zms_ate empty_ate;
+	size_t ate_size;
+	uint8_t cycle_cnt;
+	int rc = 0;
+
+	addr &= ADDR_SECT_MASK;
+
+	ate_size = zms_al_size(fs, sizeof(struct zms_ate));
+
+	LOG_DBG("Adding empty ate at %llx", (uint64_t)(addr + fs->sector_size - 2 * ate_size));
+	empty_ate.id = ZMS_HEAD_ID;
+	empty_ate.len = 0xffff;
+	empty_ate.offset = 0U;
+	empty_ate.metadata = ZMS_DEFAULT_VERSION;
+
+	rc = zms_get_sector_cycle(fs, addr, &cycle_cnt);
+	if (rc == -ENOENT) {
+		/* sector never used */
+		cycle_cnt = 0;
+	} else if (rc) {
+		/* bad flash read */
+		return rc;
+	}
+
+	/* increase cycle counter */
+	empty_ate.cycle_cnt = (cycle_cnt + 1) % BIT(8);
+	zms_ate_crc8_update(&empty_ate);
+	fs->ate_wra = addr + fs->sector_size - (2 * ate_size);
+
+	return zms_flash_ate_wrt(fs, &empty_ate);
+}
+
+static int zms_get_sector_cycle(struct zms_fs *fs, uint64_t addr, uint8_t *cycle_cnt)
+{
+	int rc;
+	size_t ate_size;
+	struct zms_ate empty_ate;
+	uint64_t empty_addr;
+
+	ate_size = zms_al_size(fs, sizeof(struct zms_ate));
+
+	empty_addr = addr & ADDR_SECT_MASK;
+	empty_addr += fs->sector_size - 2 * ate_size;
+
+	/* read the cycle counter of the current sector */
+	rc = zms_flash_ate_rd(fs, empty_addr, &empty_ate);
+	if (rc < 0) {
+		/* flash error */
+		return rc;
+	}
+
+	if (zms_empty_ate_valid(fs, &empty_ate)) {
+		*cycle_cnt = empty_ate.cycle_cnt;
+	} else {
+		/* there is no empty ATE in this sector */
+		return -ENOENT;
+	}
+
+	return 0;
```

Comment:
you can just move `return 0` to the `zms_empty_ate_valid_block` and avoid else, just return -ENOENT at the end of function.